File: cli/internal/commands/changelog.go
package commands

import (
	"fmt"
	"io"
	"strings"

	"github.com/blackcode-switzerland/bc-issues/cli/internal/client"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/config"
	"github.com/blackcode-switzerland/bc-issues/cli/internal/output"
	"github.com/spf13/cobra"
)

// defaultChangelogServer is used when the user runs `bk changelog` before ever
// logging in. The changelog is public, and the whole point of the command is to
// help an out-of-date setup get current — so it must work without a config file.
const defaultChangelogServer = "https://bc-issues.vercel.app"

// `bk changelog` mirrors GET /api/changelog: the dated record of what changed.
//
// It no longer has a --reference flag. That printed a server-hosted "Platform
// Reference" — a complete snapshot of the surface — which is precisely the kind
// of copy that drifts (its CLI version was stale before we retired it). The
// current surface is `bk guide`, embedded in THIS binary, so it can never
// describe a version you are not running. The flag is kept as a hidden alias
// that redirects, rather than vanishing into "unknown flag".
func newChangelogCmd() *cobra.Command {
	var full, reference bool
	var serverFlag string

	cmd := &cobra.Command{
		Use:         "changelog",
		Annotations: map[string]string{"routes": "GET /api/changelog"},
		Short:       "What's changed in the API and CLI (read this to get up to date)",
		Long: `Print the product changelog (GET /api/changelog).

Default: a table of dated changes, newest first.
  --full        print every dated entry in full

For how the CLI WORKS (rather than what changed), run ` + "`bk guide`" + ` — the
complete usage guide embedded in this binary.

The changelog is public, so this works even before ` + "`bk login`" + `. If a
command that used to work now fails, run ` + "`bk skill sync`" + `, then check here.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := output.Resolve(cmd)
			if err != nil {
				return err
			}

			// --reference was retired with the Platform Reference it printed.
			// Redirect instead of failing: a hint an agent can act on beats an
			// "unknown flag" it can only give up on.
			if reference {
				fmt.Fprintln(cmd.ErrOrStderr(),
					"hint: --reference was retired — the platform reference is now the embedded guide.\n"+
						"      Run `bk guide` (offline, always matches this binary).")
			}

			c := changelogClient(serverFlag)
			cl, err := c.Changelog()
			if err != nil {
				return err
			}

			return output.Render(format, cl, func(w io.Writer) error {
				if full {
					// Whole dated log, in full.
					for i, e := range cl.Entries {
						if i > 0 {
							fmt.Fprintln(w)
						}
						fmt.Fprintf(w, "## %s — %s\n\n", e.Date, e.Title)
						fmt.Fprintln(w, strings.TrimSpace(e.Markdown))
					}
					return nil
				}

				// Default: a compact table of dated entries.
				fmt.Fprintf(w, "blackcode issues — changelog (CLI latest v%s, min v%s)\n\n",
					cl.CLILatestVersion, cl.CLIMinVersion)
				tw := output.Tabwriter(w)
				fmt.Fprintln(tw, "DATE\tCHANGE")
				for _, e := range cl.Entries {
					date := e.Date
					if date == "" {
						date = "—"
					}
					fmt.Fprintf(tw, "%s\t%s\n", date, e.Title)
				}
				if err := tw.Flush(); err != nil {
					return err
				}
				fmt.Fprintln(cmd.ErrOrStderr(),
					"\nRead every entry in full: `bk changelog --full`. How the CLI works: `bk guide`.")
				return nil
			})
		},
	}

	cmd.Flags().BoolVar(&full, "full", false, "Print every dated entry in full")
	cmd.Flags().BoolVar(&reference, "reference", false, "Retired — the platform reference is now `bk guide`")
	_ = cmd.Flags().MarkDeprecated("reference", "the platform reference is now the embedded guide: run `bk guide`")
	cmd.Flags().StringVar(&serverFlag, "server", "", "Server base URL (default: your logged-in server, else "+defaultChangelogServer+")")
	return cmd
}

// changelogClient builds a client for the public changelog endpoint. It prefers
// an explicit --server, then the logged-in server from config, then the default
// host — so the command works with no config at all.
func changelogClient(serverFlag string) *client.Client {
	server := strings.TrimSpace(serverFlag)
	token := ""
	if server == "" {
		if cfg, err := config.Load(); err == nil {
			server = cfg.Server
			token = cfg.Token
		}
	}
	if server == "" {
		server = defaultChangelogServer
	}
	return client.New(server, token, "")
}

File: apps/issues/lib/changelog.ts
// Single source of truth for the product changelog, surfaced two aligned ways:
// `bk changelog` and GET /api/changelog. Both read docs/api-changelog.md — the
// dated, newest-first log of every change. Each entry is a
// `## YYYY-MM-DD — Title` section.
//
// The /changelog web page was removed on 2026-08-03: it had no human audience,
// and a page nobody reads is still a page somebody has to keep honest. The
// changelog is an AGENT surface now.
//
// It used to also serve docs/platform-reference.md, a pinned "complete snapshot
// of the API + CLI surface". That document is gone: a hand-maintained snapshot of
// a surface is a copy, and copies drift — its CLI version was already stale when
// we retired it. The current surface is now described by `bk guide`, which ships
// inside the binary and therefore always matches the binary being run.
//
// So the changelog has exactly one job now: the dated record of what changed.
// For how things WORK, an agent runs `bk guide`; for live values, `bk meta`.
//
// The file is the editable source; this module reads, parses, and renders it. It
// is bundled into the serverless output via next.config.js
// (outputFileTracingIncludes) so the reads work in production too.

import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { marked } from 'marked'
import sanitizeHtml from 'sanitize-html'
import { CLI_LATEST_VERSION, CLI_MIN_VERSION } from './cli-version'

// The changelog lives at the MONOREPO ROOT (`docs/`), not inside this app —
// architecture §7.3 makes it a platform surface that will merge one file per app
// into a single feed. This app runs with its cwd at `apps/issues`, so walk up two
// levels. `outputFileTracingRoot` + `outputFileTracingIncludes` in next.config.js
// are what carry the file into the serverless bundle; keep the two in step.
const DOCS_DIR = join(process.cwd(), '..', '..', 'docs')

export interface ChangelogEntry {
  /** ISO date `YYYY-MM-DD` parsed from the entry heading (empty if unparseable). */
  date: string
  /** The entry title (the heading text after the date). */
  title: string
  /** The entry body Markdown (without the `## heading` line). */
  markdown: string
  /** The entry body rendered to sanitized HTML. */
  html: string
}

export interface ChangelogPayload {
  /** Newest published CLI; older clients get a soft "update available" notice. */
  cli_latest_version: string
  /** Minimum CLI the API still supports; older clients are hard-blocked. */
  cli_min_version: string
  /** Every dated change, newest first (docs/api-changelog.md). */
  entries: ChangelogEntry[]
  /**
   * Where the retired Platform Reference went. Kept in the payload (rather than
   * silently dropping the field) so a client built against the old shape gets an
   * answer instead of `undefined` — the same reasoning as the 410 stubs.
   */
  reference_moved_to: string
}

// We author the changelog; this is not untrusted user input. We still sanitize
// (defense in depth) with the same tag/attribute whitelist the rich-text layer
// uses, so the rendered HTML can drop straight into a `.prose` container.
const SANITIZE_OPTS: sanitizeHtml.IOptions = {
  allowedTags: [
    'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
    'p', 'br', 'hr', 'blockquote',
    'strong', 'b', 'em', 'i', 'u', 's', 'del', 'mark', 'sub', 'sup',
    'code', 'pre',
    'ul', 'ol', 'li',
    'a', 'img',
    'table', 'thead', 'tbody', 'tr', 'th', 'td', 'caption', 'colgroup', 'col',
    'span', 'div',
  ],
  allowedAttributes: {
    a: ['href', 'title', 'target', 'rel'],
    img: ['src', 'alt', 'title'],
    td: ['colspan', 'rowspan'],
    th: ['colspan', 'rowspan'],
    '*': ['id', 'class'],
  },
  allowedSchemes: ['http', 'https', 'mailto'],
  transformTags: {
    a: sanitizeHtml.simpleTransform('a', { rel: 'noopener noreferrer', target: '_blank' }),
  },
}

// gfm for tables/strikethrough; breaks:false so the hard-wrapped prose in the
// source files renders as flowing paragraphs (not a <br> per wrapped line).
function render(md: string): string {
  const html = marked.parse(md, { async: false, gfm: true, breaks: false }) as string
  return sanitizeHtml(html, SANITIZE_OPTS)
}

// Split the dated log into entries. Everything before the first `## ` heading
// (the file title + intro) is dropped; each `## YYYY-MM-DD — Title` starts a new
// entry, and trailing `---` separators / whitespace are trimmed from each body.
function parseEntries(md: string): ChangelogEntry[] {
  const firstHeading = md.indexOf('\n## ')
  const body = firstHeading >= 0 ? md.slice(firstHeading + 1) : md
  const sections = body.split(/\n(?=## )/g)
  const entries: ChangelogEntry[] = []
  for (const section of sections) {
    const m = section.match(/^##\s+(.+?)\r?\n([\s\S]*)$/)
    if (!m) continue
    const heading = m[1].trim()
    const rest = m[2].replace(/\n+---\s*$/, '').trim()
    const dated = heading.match(/^(\d{4}-\d{2}-\d{2})\s*[—-]\s*(.+)$/)
    entries.push({
      date: dated ? dated[1] : '',
      title: (dated ? dated[2] : heading).trim(),
      markdown: rest,
      html: render(rest),
    })
  }
  return entries
}

// Read once and memoize. In production the files never change under the running
// process; in dev, editing a doc triggers a module reload which clears this.
let cached: ChangelogPayload | null = null

export function getChangelog(): ChangelogPayload {
  if (cached) return cached
  const logMd = readFileSync(join(DOCS_DIR, 'api-changelog.md'), 'utf8')
  cached = {
    cli_latest_version: CLI_LATEST_VERSION,
    cli_min_version: CLI_MIN_VERSION,
    entries: parseEntries(logMd),
    reference_moved_to: 'Run `bk guide` — the complete usage guide, embedded in the CLI binary.',
  }
  return cached
}

// The raw dated log as Markdown — used by GET /api/changelog?format=markdown and
// `bk changelog --full` so a client can grab the whole record as one document.
export function getChangelogMarkdown(): string {
  return `${readFileSync(join(DOCS_DIR, 'api-changelog.md'), 'utf8').trim()}\n`
}
